ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctAggFunctionFactory.java

```diff
@@ -18,59 +18,50 @@
 import com.google.common.collect.ImmutableList;
 import io.confluent.ksql.function.AggregateFunctionFactory;
 import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.util.DecimalUtil;
 import io.confluent.ksql.util.KsqlException;
-import java.util.Arrays;
-import java.util.HashMap;
+import io.confluent.ksql.util.SchemaUtil;
 import java.util.List;
-import java.util.Map;
 import org.apache.kafka.connect.data.Schema;
 
 public class TopkDistinctAggFunctionFactory extends AggregateFunctionFactory {
 
   private static final String NAME = "TOPKDISTINCT";
-  private final Map<Schema.Type, KsqlAggregateFunction<?, ?>> functions = new HashMap<>();
-
 
   private static final List<List<Schema>> SUPPORTED_TYPES = ImmutableList
       .<List<Schema>>builder()
       .add(ImmutableList.of(Schema.OPTIONAL_INT32_SCHEMA))
       .add(ImmutableList.of(Schema.OPTIONAL_INT64_SCHEMA))
       .add(ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA))
       .add(ImmutableList.of(Schema.OPTIONAL_STRING_SCHEMA))
+      .add(ImmutableList.of(DecimalUtil.builder(1, 1)))
       .build();
 
   public TopkDistinctAggFunctionFactory() {
     super(NAME);
-    for (KsqlAggregateFunction<?, ?> func : createDescriptionFunctions()) {
-      functions.put(((TopkDistinctKudaf) func).getOutputSchema().type(), func);
-    }
-  }
-
-  private static List<KsqlAggregateFunction<?, ?>> createDescriptionFunctions() {
-    return Arrays.asList(
-        new TopkDistinctKudaf<>(NAME, -1, 0, Schema.OPTIONAL_INT32_SCHEMA,
-            Integer.class),
-        new TopkDistinctKudaf<>(NAME, -1, 0, Schema.OPTIONAL_INT64_SCHEMA,
-            Long.class),
-        new TopkDistinctKudaf<>(NAME, -1, 0, Schema.OPTIONAL_FLOAT64_SCHEMA,
-            Double.class),
-        new TopkDistinctKudaf<>(NAME, -1, 0, Schema.OPTIONAL_STRING_SCHEMA,
-            String.class)
-    );
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public KsqlAggregateFunction getProperAggregateFunction(final List<Schema> argTypeList) {
     if (argTypeList.isEmpty()) {
       throw new KsqlException("TOPKDISTINCT function should have two arguments.");
     }
 
-    final KsqlAggregateFunction<?, ?> function = functions.get(argTypeList.get(0).type());
-    if (function == null) {
-      throw new KsqlException("No TOPKDISTINCT aggregate function with " + argTypeList.get(0)
-          + " argument type exists!");
+    final Schema argSchema = argTypeList.get(0);
+    switch (argSchema.type()) {
+      case INT32:
+      case INT64:
+      case FLOAT64:
+      case STRING:
+        return new TopkDistinctKudaf(NAME, 0, -1, argSchema, SchemaUtil.getJavaType(argSchema));
+      case BYTES:
```

Comment:
I see you've added query validation tests for this, but perhaps also worth adding a `DecimalTopkDistinctKudafTest.java` unit test file, analogous to the existing ones for the other types?

Reply:
:scratches-head: I was sure I had done that!